driver/adfs_proxy.cc

```diff
@@ -28,31 +28,245 @@
 // http://www.gnu.org/licenses/gpl-2.0.html.
 
 #include "adfs_proxy.h"
+#include <regex>
 #include "driver.h"
 
+#define SIGN_IN_PAGE_URL "/adfs/ls/IdpInitiatedSignOn.aspx?loginToRp=urn:amazon:webservices"
+
+std::unordered_map<std::string, TOKEN_INFO> ADFS_PROXY::token_cache;
+std::mutex ADFS_PROXY::token_cache_mutex;
+
 ADFS_PROXY::ADFS_PROXY(DBC* dbc, DataSource* ds) : ADFS_PROXY(dbc, ds, nullptr) {};
 
 ADFS_PROXY::ADFS_PROXY(DBC* dbc, DataSource* ds, CONNECTION_PROXY* next_proxy) : CONNECTION_PROXY(dbc, ds) {
   this->next_proxy = next_proxy;
-  if (ds->opt_AUTH_REGION) {
-      this->auth_util = std::make_shared<AUTH_UTIL>((const char*)ds->opt_AUTH_REGION);
+  std::string host{static_cast<const char*>(ds->opt_IDP_ENDPOINT)};
+  host += ":" + std::to_string(ds->opt_IDP_PORT);
+
+  const int client_connect_timeout = ds->opt_CLIENT_CONNECT_TIMEOUT;
+  const int client_socket_timeout = ds->opt_CLIENT_SOCKET_TIMEOUT;
+  const bool enable_ssl = ds->opt_ENABLE_SSL;
+  this->saml_util = std::make_shared<ADFS_SAML_UTIL>(host, client_connect_timeout, client_socket_timeout, enable_ssl);
+}
+
+void ADFS_PROXY::clear_token_cache() {
+  std::unique_lock<std::mutex> lock(token_cache_mutex);
+  token_cache.clear();
+}
+
+ADFS_SAML_UTIL::ADFS_SAML_UTIL(const std::shared_ptr<SAML_HTTP_CLIENT>& client) { this->http_client = client; }
+
+ADFS_SAML_UTIL::ADFS_SAML_UTIL(std::string host, int connect_timeout, int socket_timeout, bool enable_ssl) {
+  this->http_client =
+      std::make_shared<SAML_HTTP_CLIENT>("https://" + host, connect_timeout, socket_timeout, enable_ssl);
+}
+
+std::string ADFS_SAML_UTIL::get_saml_assertion(DataSource* ds) {
+  nlohmann::json res;
+  try {
+    res = this->http_client->get(std::string(SIGN_IN_PAGE_URL));
+  } catch (SAML_HTTP_EXCEPTION& e) {
+    const std::string error =
+        "Failed to get sign-in page from ADFS: " + e.error_message() + ". Please verify your IDP endpoint.";
+    throw SAML_HTTP_EXCEPTION(error);
+  }
+
+  const auto body = std::string(res);
+  std::smatch m;
+  if (!std::regex_search(body, m, ADFS_REGEX::FORM_ACTION_PATTERN)) {
+    return std::string();
   }
-  else {
-      this->auth_util = std::make_shared<AUTH_UTIL>();
+  std::string form_action = escape_html_entity(m.str(1));
+  const std::string params = get_parameters_from_html(ds, body);
+  const std::string content = get_form_action_body(form_action, params);
+  printf("hjello %s\n", content.c_str());
+  if (std::regex_search(content, m, ADFS_REGEX::SAML_RESPONSE_PATTERN)) {
+    return m.str(1);
   }
+  return std::string();
+}
+
+std::string ADFS_SAML_UTIL::escape_html_entity(const std::string& html) {
```

Comment:
Very minor: It looks like this function is actually "un-escaping" an HTML entity (i.e. going from an escaped HTML entity like `"&lt;"` to `'<'`. Perhaps this method name should reflect that?